File: algorithms/fcfs/cpp/stl/fcfs-sort.cpp
#include <iostream>
#include <vector>
#include <algorithm>

struct Process {
	unsigned int pid;
	unsigned int at;
	unsigned int bt;
	unsigned int ct;
};

// ------------------------------------------------------
// Finding the waiting and turn around time, after computing the
// completion time.

void findCompletionTime(std::vector<Process> &proc) {
	for(auto &p: proc)
		if (p.at > (&p - 1)->ct)
			p.ct = p.at + p.bt;
		else
			p.ct = (&p - 1)->ct + p.bt;
}

void findAvgTime(std::vector<Process> &proc) {
	std::sort(
			proc.begin(),
			proc.end(),
			[](const Process &a, const Process &b) {
			return a.at < b.at;
			}
		 );
	findCompletionTime(proc);
	std::sort(
			proc.begin(),
			proc.end(),
			[](const Process &a, const Process &b) {
			return a.pid < b.pid;
			}
		 );

	float ta{0}, w{0};

	std::cout << "PID\tAT\tBT\tCT\tTAT\tWT\n";
	for (const auto &p: proc) {
		unsigned pTA = p.ct - p.at;
		unsigned pW = pTA - p.bt;
		std::cout << p.pid << '\t' << p.at << '\t'
			<< p.bt << '\t' << p.ct << '\t'
			<< pTA << '\t' << pW << '\n';
		ta += pTA;
		w += pW;
	}

	ta /= proc.size();
	w /= proc.size();

	std::cout << std::endl;
	std::cout << "Average Waiting Time: " << w << '\n' 
		<< "Average Turn-Around Time: " << ta << '\n';
}

// ------------------------------------------------------
// Driver for the program

int main() {
	std::vector<Process> proc = {
		{0, 5, 2, 0},
		{1, 5, 3, 0},
		{2, 1, 4, 0},
		{3, 2, 2, 0},
		{4, 1, 2, 0},
		{5, 1, 1, 0}
	};

	findAvgTime(proc);

	std::cout << '\n';
	return 0;
}

File: algorithms/fcfs/fcfs-sort.java
import java.util.Arrays;
import java.util.Collections;

class Process {
    int pid, at, bt, ct;

    public Process(int pid, int at, int bt) {
        this.pid = pid;
        this.at = at;
        this.bt = bt;
        this.ct = 0;
    }
}

class fcfs {
    public static void main(String[] args) {
        Process proc[] = new Process[] { new Process(0, 1, 5), new Process(1, 0, 5), new Process(2, 0, 2),
                new Process(3, 2, 4) };
        float ta = 0, w = 0;

        Collections.sort(Arrays.asList(proc), (Process a, Process b) -> {
            if (a.at < b.at)
                return -1;
            return 1;
        });
        for (int i = 0; i < proc.length; i++) {
            if (i == 0 || proc[i - 1].ct < proc[i].at)
                proc[i].ct = proc[i].at;
            else
                proc[i].ct = proc[i - 1].ct;
            proc[i].ct += +proc[i].bt;
            ta += proc[i].ct - proc[i].at;
            w += proc[i].ct - (proc[i].at + proc[i].bt);
        }
        Collections.sort(Arrays.asList(proc), (Process a, Process b) -> {
            if (a.pid < b.pid)
                return -1;
            return 1;
        });

        System.out.println("PID\tAT\tBT\tCT\tTAT\tWT");
        for (Process p : proc) {
            System.out.println(p.pid + "\t" + p.at + '\t' + p.bt + '\t' + p.ct + '\t' + (p.ct - p.at) + '\t'
                    + (p.ct - (p.at + p.bt)));
        }
        System.out.println("Avg. TAT: " + ta / proc.length);
        System.out.println("Avg. WT: " + w / proc.length);
    }
}